concordium-std/src/state_btree.rs

```diff
@@ -0,0 +1,1481 @@
+    fn copy_then_increment(&mut self) -> Self {
+        let current = *self;
+        self.id += 1;
+        current
+    }
+
+    /// Construct the key for the node in the key-value store from the node ID.
+    fn as_key(&self, prefix: &StateItemPrefix) -> [u8; BTREE_NODE_KEY_SIZE] {
+        // Create an uninitialized array of `MaybeUninit`. The `assume_init` is
+        // safe because the type we are claiming to have initialized here is a
+        // bunch of `MaybeUninit`s, which do not require initialization.
+        let mut prefixed: [mem::MaybeUninit<u8>; BTREE_NODE_KEY_SIZE] =
+            unsafe { mem::MaybeUninit::uninit().assume_init() };
+        for i in 0..STATE_ITEM_PREFIX_SIZE {
+            prefixed[i].write(prefix[i]);
+        }
+        let id_bytes = self.id.to_le_bytes();
+        for i in 0..id_bytes.len() {
+            prefixed[STATE_ITEM_PREFIX_SIZE + i].write(id_bytes[i]);
+        }
+        // Transmuting away the maybeuninit is safe since we have initialized all of
+        // them.
+        unsafe { mem::transmute(prefixed) }
+    }
+}
+
+impl<const M: usize, K, V, S> Serial for StateBTreeMap<K, V, S, M> {
+    fn serial<W: Write>(&self, out: &mut W) -> Result<(), W::Err> {
+        self.map.serial(out)?;
+        self.ordered_set.serial(out)
+    }
+}
+
+impl<const M: usize, K, V, S: HasStateApi> DeserialWithState<S> for StateBTreeMap<K, V, S, M> {
+    fn deserial_with_state<R: Read>(state: &S, source: &mut R) -> ParseResult<Self> {
+        let map = DeserialWithState::deserial_with_state(state, source)?;
+        let ordered_set = DeserialWithState::deserial_with_state(state, source)?;
+        Ok(Self {
+            map,
+            ordered_set,
+        })
+    }
+}
+
+impl<const M: usize, K, S> Serial for StateBTreeSet<K, S, M> {
+    fn serial<W: Write>(&self, out: &mut W) -> Result<(), W::Err> {
+        self.prefix.serial(out)?;
+        self.root.serial(out)?;
+        self.len.serial(out)?;
+        self.next_node_id.serial(out)
+    }
+}
+
+impl<const M: usize, K, S: HasStateApi> DeserialWithState<S> for StateBTreeSet<K, S, M> {
+    fn deserial_with_state<R: Read>(state: &S, source: &mut R) -> ParseResult<Self> {
+        let prefix = source.get()?;
+        let root = source.get()?;
+        let len = source.get()?;
+        let next_node_id = source.get()?;
+
+        Ok(Self {
+            _marker_key: Default::default(),
+            prefix,
+            state_api: state.clone(),
+            root,
+            len,
+            next_node_id,
+        })
+    }
+}
+
+impl<const M: usize, K: Serial> Serial for Node<M, K> {
+    fn serial<W: Write>(&self, out: &mut W) -> Result<(), W::Err> {
+        self.keys.serial(out)?;
+        self.children.serial(out)
+    }
+}
+
+impl<const M: usize, K: Deserial> Deserial for Node<M, K> {
+    fn deserial<R: Read>(source: &mut R) -> ParseResult<Self> {
+        let keys = source.get()?;
+        let children = source.get()?;
```

Comment:
Won't this use u64 for length? Can't we use less?

Reply:
Also why not just derive it.

YOu can use derive macros as long as you import `self as concordium_std`.

Reply:
It uses u32, since `Serial` in `concordium_contracts_common` defaults to `u32` for length.
However, `Serial` in `concordium-base` defaults to `u64`.